From c72066f1cd4bf28eb3e7de9acb30f730e60fb6e1 Mon Sep 17 00:00:00 2001
From: gouse <gouse@qti.qualcomm.com>
Date: Tue, 10 Mar 2026 16:18:00 +0530
Subject: [PATCH] 645-iw-Update-NL80211-commands-and-attributes

This patch updates the iw utility to support newly added NL80211
commands and attributes related to MU‑EDCA configuration. The changes
extend the userspace interface so that MU‑EDCA parameters can be set
through cfg80211 using the standard nl80211 command set.

Feature: EDCA

Patch-dependency: None
Patch-work: None

Signed-off-by: gouse <gouse@qti.qualcomm.com>
---
 nl80211.h | 485 +++++++++++++++++++++++++++++++++++++++++++++++++++++-
 1 file changed, 476 insertions(+), 9 deletions(-)

--- a/nl80211.h
+++ b/nl80211.h
@@ -3001,6 +3001,26 @@ enum nl80211_commands {
  * @NL80211_ATTR_6GHZ_DEVICE_DEPLOYMENT_TYPE: Attribute that denotes the 6 GHz
  *	device deployment type.
  *
+ * @NL80211_ATTR_CONTROL_MIC_PAD: Mic padding for control frame protection.
+ *
+ * @NL80211_ATTR_USE_CFP: Indicate the use of control frame protection by station.
+ *
+ * @NL80211_ATTR_CFP: Indicate whether the particular vdev supports control
+ *      frame protection
+ *
+ * @NL80211_ATTR_EXT_MLD_CAPA_AND_OPS: Extended MLD Capabilities and Operations
+ *      (u16)
+ *
+ * @NL80211_ATTR_BITRATE: This attribute is used with %NL80211_CMD_FRAME to
+ *     send legacy bitrate information of management packets to userspace.
+ *
+ * * @NL80211_ATTR_UHR_CAPABILITY: UHR Capability information element (from
+ *      association request when used with NL80211_CMD_NEW_STATION). Can be set
+ *      only if HE/EHT are also available.
+ * @NL80211_ATTR_DISABLE_UHR: Force UHR capable interfaces to disable
+ *      this feature during association. This is a flag attribute.
+ *      Currently only supported in mac80211 drivers.
+
  * @NL80211_ATTR_PCIE: Nested attributes associated with PCIe low power and
  *	gen/lane mode transitions. See &enum nl80211_pcie_attrs
  *
@@ -3010,6 +3030,9 @@ enum nl80211_commands {
  * @NL80211_ATTR_DPS_ASSIST: flag to enable/disable DPS assisting role. See
  *	&enum nl80211_dps_assist_attrs
  *
+ * @NL80211_ATTR_HE_MUEDCA_MODE: Attribute that denotes the mode of setting MU
+ *	EDCA parameters(See enum nl80211_muedca_mode).
+ *
  * @NUM_NL80211_ATTR: total number of nl80211_attrs available
  * @NL80211_ATTR_MAX: highest attribute number currently defined
  * @__NL80211_ATTR_AFTER_LAST: internal use
@@ -3612,9 +3635,25 @@ enum nl80211_attrs {
 
 	NL80211_ATTR_6GHZ_DEVICE_DEPLOYMENT_TYPE,
 
+	NL80211_ATTR_CONTROL_MIC_PAD,
+
+	NL80211_ATTR_USE_CFP,
+
+	NL80211_ATTR_CFP,
+
+	NL80211_ATTR_CIGTK,
+
+	NL80211_ATTR_EXT_MLD_CAPA_AND_OPS,
+
+	NL80211_ATTR_BITRATE,
+
+	NL80211_ATTR_UHR_CAPABILITY,
+	NL80211_ATTR_DISABLE_UHR,
 	NL80211_ATTR_PCIE,
 	NL80211_ATTR_DCVS,
 	NL80211_ATTR_DPS_ASSIST,
+
+	NL80211_ATTR_HE_MUEDCA_MODE,
 	/* add attributes here, update the policy in nl80211.c */
 
 	__NL80211_ATTR_AFTER_LAST,
